### ISSUE TITLE
New c4doc_update implementation leads to use after free

### ISSUE THREAD

**Issue by borrrden** [couchbase/couchbase-lite-core#487]
The implementation of c4doc_update introduced in commit d375c8908943477a45fa2c1feb3039001b31c990 to address #478 comes with some side effects.  Since it doesn't copy the document like the alternate version does, then calling c4doc_update, then freeing the original `C4Document` leads to the inability of arrays and dictionaries retrieved from that document to retrieve their data since it is already freed.  This can happen via the following CBL steps:

1. Create an array (or dictionary)
2. Set it into a document
3. Save the document
4. Retrieve the document
5. Retrieve the collection from the document
6. Delete the document
7. Try to read data out of the collection

If it is fast enough the issue won't be seen on Objective-C because the document will be kept alive slightly longer by the autorelease pool, but on .NET it happens consistently and CBL Objective-C can be modified to demonstrate the behavior via the following patch so that it happens consistently every time the test titled `testArrayAfterDeleteDocument` is run:

[patch.txt](https://github.com/couchbase/couchbase-lite-core/files/1837439/patch.txt) (renamed patch.diff)

**Comment by snej** [comment]
Those steps make an assumption about undocumented behavior of c4doc_update: that the new C4Document it returns contains the same revision body pointers (`alloc_slice`s) as the input C4Document, so that existing Fleece pointers into the revision bodies will still be valid after the input document is freed.

This was true of the old implementation, but not of the new one.

The same test passes on Objective-C, but the path to get there is very complicated and goes through layers of “mutable fleece.”

>Objective-C can be modified to demonstrate the behavior via the following patch

The patch forces the old CBLC4Document object to free its C4Document. But that old CBLC4Document is still referenced by the mutable-Fleece objects used by CBLDictionary and CBLArray objects (via their DocContext), so it pulls the rug out from under them.

In other words, the reason this bug doesn't happen in Obj-C is that the mutable-Fleece objects ensure that the C4Document owning their memory stays around, even if it's no longer referenced by the CBLDocument.

**Comment by snej** [comment]
I could change `c4doc_update` to preserve the `alloc_slice`s of the old revision bodies, but it sounds like @borrrden already has a workaround that I think uses the same approach as Obj-C.

**Comment by snej** [comment]
Freeing a C4Document invalidates all of the memory it points to -- docID, revIDs, revision bodies.

**Comment by snej** [comment]
I've been thinking it'd be good to expose alloc_slice in the C API somehow, i.e. a form of C4Slice that you can retain and release. That would give CBL finer grained control over memory management.

**Comment by borrrden** [comment]
Nope, just whether or not to implement what you said (tracked elsewhere)
